### ISSUE TITLE
slp-vectorizer pass removes calls to zext/trunc which breaks correctness

### ISSUE THREAD

**Issue by jabraham17** [llvm/llvm-project#118410]
The slp-vectorizer in LLVM 19 seems to erroneously remove calls to `zext`/`trunc`, which breaks the correctness of the code. This is a regression from LLVM 18 to 19.

This [compiler explorer link](https://godbolt.org/z/zof5rrxze) shows the issue, I will also paste the code here.

Input LLVM IR
```llvm
target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128-Fn32"
target triple = "arm64-apple-macosx13.0.0"

define i64 @chpl_gen_main(i16 %0) {
chpl__init_repro.exit:
  %1 = zext i16 %0 to i64
  %2 = or i64 0, %1
  %3 = tail call i64 @llvm.abs.i64(i64 %2, i1 true)
  %4 = trunc i64 %3 to i16
  %5 = zext i16 0 to i64
  %6 = or i64 0, %5
  %7 = tail call i64 @llvm.abs.i64(i64 %6, i1 true)
  %8 = trunc i64 %7 to i16
  %9 = or i16 %8, %4
  %10 = zext i16 0 to i64
  %11 = or i64 0, %10
  %12 = tail call i64 @llvm.abs.i64(i64 %11, i1 true)
  %13 = trunc i64 %12 to i16
  %14 = or i16 %9, %13
  %15 = zext i16 0 to i64
  %16 = or i64 0, %15
  %17 = tail call i64 @llvm.abs.i64(i64 %16, i1 true)
  %18 = trunc i64 %17 to i16
  %19 = or i16 %14, %18
  %20 = zext i16 0 to i64
  %21 = or i64 0, %20
  %22 = tail call i64 @llvm.abs.i64(i64 %21, i1 true)
  %23 = trunc i64 %22 to i16
  %24 = or i16 %19, %23
  %25 = zext i16 0 to i64
  %26 = or i64 0, %25
  %27 = tail call i64 @llvm.abs.i64(i64 %26, i1 true)
  %28 = trunc i64 %27 to i16
  %29 = or i16 %24, %28
  %30 = zext i16 0 to i64
  %31 = or i64 %30, 0
  %32 = tail call i64 @llvm.abs.i64(i64 %31, i1 true)
  %33 = trunc i64 %32 to i16
  %34 = or i16 %29, %33
  %35 = zext i16 0 to i64
  %36 = or i64 0, %35
  %37 = tail call i64 @llvm.abs.i64(i64 %36, i1 true)
  %38 = trunc i64 %37 to i16
  %39 = or i16 %34, %38
  %40 = zext i16 0 to i64
  %41 = or i64 0, %40
  %42 = tail call i64 @llvm.abs.i64(i64 %41, i1 true)
  %43 = trunc i64 %42 to i16
  %44 = or i16 %39, %43
  %45 = zext i16 0 to i64
  %46 = or i64 %45, 0
  %47 = tail call i64 @llvm.abs.i64(i64 %46, i1 true)
  %48 = trunc i64 %47 to i16
  %49 = or i16 %44, %48
  %50 = zext i16 0 to i64
  %51 = or i64 0, %50
  %52 = tail call i64 @llvm.abs.i64(i64 %51, i1 true)
  %53 = trunc i64 %52 to i16
  %54 = or i16 %49, %53
  %55 = zext i16 0 to i64
  %56 = or i64 0, %55
  %57 = tail call i64 @llvm.abs.i64(i64 %56, i1 true)
  %58 = trunc i64 %57 to i16
  %59 = or i16 %54, %58
  %60 = zext i16 0 to i64
  %61 = or i64 %60, 0
  %62 = tail call i64 @llvm.abs.i64(i64 %61, i1 true)
  %63 = trunc i64 %62 to i16
  %64 = or i16 %59, %63
  %65 = zext i16 0 to i64
  %66 = or i64 0, %65
  %67 = tail call i64 @llvm.abs.i64(i64 %66, i1 true)
  %68 = trunc i64 %67 to i16
  %69 = or i16 %64, %68
  %70 = zext i16 0 to i64
  %71 = or i64 0, %70
  %72 = tail call i64 @llvm.abs.i64(i64 %71, i1 true)
  %73 = trunc i64 %72 to i16
  %74 = or i16 %69, %73
  %75 = zext i16 0 to i64
  %76 = or i64 %75, 0
  %77 = tail call i64 @llvm.abs.i64(i64 %76, i1 true)
  %78 = trunc i64 %77 to i16
  %79 = or i16 %74, %78
  store i16 %79, ptr null, align 2
  ret i64 0
}

; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
declare i64 @llvm.abs.i64(i64, i1 immarg) #0

; uselistorder directives
uselistorder ptr @llvm.abs.i64, { 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0 }

attributes #0 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }
```

Running `opt --passes='slp-vectorizer'` with LLVM 18
```llvm
source_filename = "/app/example.ll"
target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128-Fn32"
target triple = "arm64-apple-macosx13.0.0"

define i64 @chpl_gen_main(i16 %0) {
  %1 = insertelement <16 x i16> <i16 poison, i16 0, i16 0, i16 0, i16 0, i16 0, i16 0, i16 0, i16 0, i16 0, i16 0, i16 0, i16 0, i16 0, i16 0, i16 0>, i16 %0, i32 0
  %2 = zext <16 x i16> %1 to <16 x i64>
  %3 = or <16 x i64> %2, zeroinitializer
  %4 = call <16 x i64> @llvm.abs.v16i64(<16 x i64> %3, i1 true)
  %5 = trunc <16 x i64> %4 to <16 x i16>
  %6 = call i16 @llvm.vector.reduce.or.v16i16(<16 x i16> %5)
  store i16 %6, ptr null, align 2
  ret i64 0
}

declare i64 @llvm.abs.i64(i64, i1 immarg) #0

declare <16 x i64> @llvm.abs.v16i64(<16 x i64>, i1 immarg) #0

declare i16 @llvm.vector.reduce.or.v16i16(<16 x i16>) #0

attributes #0 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }
```

Running `opt --passes='slp-vectorizer'` with LLVM 19
```llvm
source_filename = "/app/example.ll"
target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128-Fn32"
target triple = "arm64-apple-macosx13.0.0"

define i64 @chpl_gen_main(i16 %0) {
  %1 = insertelement <16 x i16> <i16 0, i16 poison, i16 0, i16 0, i16 0, i16 0, i16 0, i16 0, i16 0, i16 0, i16 0, i16 0, i16 0, i16 0, i16 0, i16 0>, i16 %0, i32 1
  %2 = or <16 x i16> %1, zeroinitializer
  %3 = call <16 x i16> @llvm.abs.v16i16(<16 x i16> %2, i1 false)
  %4 = call i16 @llvm.vector.reduce.or.v16i16(<16 x i16> %3)
  store i16 %4, ptr null, align 2
  ret i64 0
}

declare i64 @llvm.abs.i64(i64, i1 immarg) #0

declare <16 x i16> @llvm.abs.v16i16(<16 x i16>, i1 immarg) #0

declare i16 @llvm.vector.reduce.or.v16i16(<16 x i16>) #0

attributes #0 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }
```

The original high level code is written in Chapel, and amounts to `(uint16)abs((int64)a - (int64)b)`, where `a` and `b` are both `uint16`. For more details on the Chapel code, see [this issue](https://github.com/chapel-lang/chapel/issues/26301)

**Comment by jabraham17** [comment]
I think this may be a duplicate of https://github.com/llvm/llvm-project/issues/112577 which is fixed on main and will be in LLVM 19.1.5?

**Comment by alexey-bataev** [comment]
> I think this may be a duplicate of [#112577](https://github.com/llvm/llvm-project/issues/112577) which is fixed on main and will be in LLVM 19.1.5?

Yes, it is duplicate of the previous bug, which is fixed already
